controllers/utils/multicloud_util.go

```diff
@@ -0,0 +1,306 @@
+package utils
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	actionv1beta1 "github.com/open-cluster-management/multicloud-operators-foundation/pkg/apis/action/v1beta1"
+	viewv1beta1 "github.com/open-cluster-management/multicloud-operators-foundation/pkg/apis/view/v1beta1"
+	ranv1alpha1 "github.com/openshift-kni/cluster-group-upgrades-operator/api/v1alpha1"
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+)
+
+var multiCloudLog = ctrl.Log.WithName("multiCloudLog")
+
+// ProcessSubscriptionManagedClusterView processes the content of a view that is configured to watch a Subscription
+// type object and takes the necessary actions to approve the InstallPlan associated with that Subscription.
+func ProcessSubscriptionManagedClusterView(
+	ctx context.Context, c client.Client, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade,
+	clusterName string, mcv *viewv1beta1.ManagedClusterView) (int, error) {
+
+	conditionMCVforSub := meta.FindStatusCondition(mcv.Status.Conditions, viewv1beta1.ConditionViewProcessing)
+	if conditionMCVforSub == nil {
+		multiCloudLog.Info("ManagedClusterView was not (yet) ready, trying again later",
+			"managedcluserview", mcv.ObjectMeta.Name, "namespace", mcv.ObjectMeta.Namespace)
+		return MultiCloudPendingStatus, nil
+	}
+
+	if conditionMCVforSub.Reason != viewv1beta1.ReasonGetResource {
+		multiCloudLog.Info("ManagedClusterView was not able to retrieve the requested resource (yet), trying again later",
+			"managedclusterview", mcv.ObjectMeta.Name, "namespace", mcv.ObjectMeta.Namespace)
+		return MultiCloudPendingStatus, nil
+	}
+
+	// Check that the ManagedClusterView was able to retrieve the information on the subscription.
+	// We do this by checking it's status and reason.
+	if conditionMCVforSub.Status == "True" && conditionMCVforSub.Reason == viewv1beta1.ReasonGetResource {
+		// Get the subscription content from the ManagedClusterView.
+		subscription := operatorsv1alpha1.Subscription{}
+		json.Unmarshal([]byte(mcv.Status.Result.Raw), &subscription)
+
+		// If the subscription's status is "UpgradePending" approve the installPlan. For any other value of the state, continue.
+		if subscription.Status.State != SubscriptionStateUpgradePending {
+			multiCloudLog.Info("Subscription State is not pending upgrade",
+				"subscription", subscription.ObjectMeta.Name, "namespace", subscription.ObjectMeta.Namespace)
```

Comment:
In a subsequent cleanup pass, this message may be slightly confusing to a user in the case where a subscription is included in the managedPolicies list but is already in the "AtLatestKnown" state. In that case we should either print nothing (since there is nothing for us to do) or change the Info log text to something like "Subscription is already at latest known".